#### (ASHFAQUE AZAD MSc Computer Science (Data Analytics))

In [1]:
import numpy as np
import pandas as pd
df1 = pd.read_csv("dataframe01.csv")
df2 = pd.read_csv("dataframe02.csv")
df3 = pd.read_csv("dataframe03.csv")
df4 = pd.read_csv("dataframe04.csv")
df5 = pd.read_csv("dataframe05.csv")
df6 = pd.read_csv("dataframe06.csv")

In [2]:
del df1['Unnamed: 0']
del df2['Unnamed: 0']
del df3['Unnamed: 0']
del df4['Unnamed: 0']
del df5['Unnamed: 0']
del df6['Unnamed: 0']

In [3]:
dataset=pd.concat([df1,df2,df3,df4,df5,df6], axis=0)

In [4]:
dataset=dataset.reset_index(drop=True)
#0 =original image
#1 =median filter on the original image
#2 =adversarial inputs
#3 =defense

In [5]:
dataset.to_csv("final_datset.csv")

In [6]:

dataset=dataset.replace("*",2000)

## Original Images correctly classified after median filtering

In [7]:
# % of original rightly classified using median filter 3x3
np.sum(dataset['0'].astype(int)==dataset['1'].astype(int))/len(dataset)*100

90.0

## Percentage of successfully generated adversarial examples which fools 2 pre-trained nets

In [8]:
# % of successfully targeted adversarial of class 390
np.sum(dataset['2'].astype(int)==390)/len(dataset)*100

65.83333333333333

## Percentage of images successfully defended by the model

In [9]:
# % of correctly classified images after defense
np.sum(dataset['0'].astype(int)==dataset['3'].astype(int))/len(dataset)*100

87.5

In [13]:
# np.sum(dataset['1'].astype(int)==dataset['3'].astype(int))/len(dataset)*100

In [10]:
import pandas as pd
q = pd.read_csv("final_datset.csv")
del q['Unnamed: 0']
q390=q[q['2']=='390']#subset of the dataset where the adversarial successfully fooled the network
# % of original rightly classified by the ensemble network
import numpy as np
q390=q390.replace("*",2000)
np.sum(q390['0'].astype(int)==q390['3'].astype(int))/len(q390)*100

82.27848101265823

In [11]:
import pandas as pd
q = pd.read_csv("final_datset.csv")
del q['Unnamed: 0']
q390=q[q['2']!='390']#subset of the dataset where the adversarial didn't fool
# % of original rightly classified by the ensemble network
import numpy as np
q390=q390.replace("*",2000)
np.sum(q390['0'].astype(int)==q390['3'].astype(int))/len(q390)*100

97.5609756097561

# The model was able to defend against the attack 87.5% of times.
## Of the 65% successful generation of adversarial images, the model was successful in defending against them 82.2% of the times